### DISCUSSION TITLE
Nested clone/synchronize

### DISCUSSION THREAD

**Discussion by yinchi** [r-simmer/simmer#292] [Q&A] [answered]
This is the code I wrote to test nested clone/synchronize:

```
traj <- trajectory() |>
  clone(3) |>
  log_("Clone level 1") |>
  clone(3) |>
  log_("Clone level 2") |>
  synchronize() |>
  log_("Back to level 1") |>
  synchronize() |>
  log_("Done")
env <- simmer() |>
  add_generator("Clone test", traj, at(0))
env |> run()
```

"Clone level 1" and "Clone level 2" appear 3 and 9 times in the result, as expected, but "Back to level 1" only appears once.  How can I get the inner synchronize() to only apply to the inner clone()?

**Comment by Enchufa2** [comment, marked as answer]
There is no concept of "inner" or "outer". Calling clone several times for the same arrival (or clones of it) just creates more clones, so `synchronize()` synchronizes all of them.

**Reply by yinchi** [reply]
So if I wanted to:

1. Break an object into medium pieces
2. Break each medium piece into small pieces
3. Do work on each small piece
4. Rejoin the small pieces into the medium pieces they came from
5. Do additional work on each medium piece
6. Rejoin the finished medium pieces into the original object

Is there a way to do so?  Sorry for the repeated messages, but this pathway is critical to a project I am doing.

**Reply by Enchufa2** [reply]
I would generate the small pieces in the first place and then batch/separate as needed in different stages.

**Reply by Enchufa2** [reply]
Or, if you only need to keep track of the object, you could 

1. Break into small pieces
2. Do work
3. Sync
4. Break into medium pieces
5. Do work
6. Sync

**Comment by yinchi** [comment]
I see.  Is there any way of getting the behaviour I want? The following hits each `log_()` the correct number of times, but set attributes are lost when batching:

```
traj_inner <- \(n) trajectory() |>
  set_attribute("n", n) |> 
  log_(\() paste("Clone level 1, n =", get_attribute(env,"n"))) |> 
  clone(3) |>
  log_(\() paste("Clone level 2, n =", get_attribute(env,"n"))) |> 
  timeout(1) |>
  batch(3, permanent = T, rule = \() get_attribute(env,"n") == n, name = "inner") |>
  log_(\() paste("Back to level 1, n =", get_attribute(env,"n")))

traj_outer <- trajectory("Clone/batch test") |>
  clone(3, lapply(1:3, \(n) traj_inner(n))) |>
  batch(3, permanent = T, name = "top_level") |>
  log_("Done")

env <- simmer() |>
  add_generator("Task", traj_outer, at(0))
env |> run()
```

If I want to copy over all attributes from the first (or last) arrival to enter a batch, can that be done?

**Reply by Enchufa2** [reply]
Batch attributes are transferred to arrivals, but not the other way around. Otherwise, which ones should be copied, what if there are conflicts? You could use global attributes for that.
